### PR TITLE
Update build command to automatically calculate the total and return collateral values

### PR DESCRIPTION
**Pull request by Jimbo4350**
The build command will now automatically calculate these values and returns the excess collateral to the change address
Resolves: #4075

### CODE REVIEW

File: cardano-api/src/Cardano/Api/Fees.hs [hunk 1]
```diff
@@ -1019,10 +1031,92 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
...
+            -- collateral tx inputs to cover the tx
+            let txOuts = catMaybes [ Map.lookup txin utxo' | txin <- collIns]
+                totalCollateralLovelace = mconcat $ map (\(TxOut _ txOutVal _ _) -> txOutValueToLovelace txOutVal) txOuts
+                requiredCollateral@(Lovelace reqAmt) = fromIntegral colPerc * fee
+                totalCollateral = TxTotalCollateral retColSup . fromShelleyLovelace
+                                                              . Ledger.rationalToCoinViaCeiling
+                                                              $ reqAmt % 100
+                enoughCollateral = totalCollateralLovelace * 100 >= requiredCollateral
+                Lovelace amt = totalCollateralLovelace * 100 - requiredCollateral
+                returnCollateral = fromShelleyLovelace . Ledger.rationalToCoinViaCeiling $ amt % 100
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/Fees.hs:1093]: What's the `100` factor for?

File: cardano-api/src/Cardano/Api/Fees.hs [hunk 2]
```diff
@@ -1019,10 +1031,92 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
...
+                                 (TxOut cAddr (lovelaceToTxOutValue $ Lovelace (2^(64 :: Integer)) - 1)
+                                 TxOutDatumNone ReferenceScriptNone)
+                 dummyTotCol = TxTotalCollateral retColSup (Lovelace (2^(32 :: Integer) - 1))
+             in case (txReturnCollateral, txTotalCollateral) of
+                  (rc@TxReturnCollateral{}, tc@TxTotalCollateral{}) -> (rc, tc)
+                  (rc@TxReturnCollateral{},TxTotalCollateralNone) -> (rc, dummyTotCol)
+                  (TxReturnCollateralNone,tc@TxTotalCollateral{}) -> (dummyRetCol, tc)
+                  (TxReturnCollateralNone, TxTotalCollateralNone) -> (dummyRetCol, dummyTotCol)
+
+   calcReturnAndTotalCollateral
```
Comment (newhoggy) [inline cardano-api/src/Cardano/Api/Fees.hs:1063]: Please add docs and links to any supporting documentation.

**Comment by newhoggy** [comment]
Can you paste/attach the commands/files you used to test?

**Comment by Jimbo4350** [comment]
> Can you paste/attach the commands/files you used to test?

Removing the fields:
```
  --tx-in-collateral "$txinCollateral" \
  --tx-total-collateral 529503 \
```
from the shell script `scripts/babbage/example-babbage-script-usage.sh` is a way to test this functionality is working via a failing plutus script. We need to update the cli's tx pretty printing to include total and return collateral fields. This would be another way to inspect that the automatic balancing is working.

**Review by saratomaz — APPROVED** [review]
The PR pass in all our regression tests related to plutus V2.

Also new tests was added to comprove that the collateral amount charged are accorging to the collateral percentage present in protocol parameters and that the change address is now the default address to return the excess of the collaterals amount.

_Related issue on cardano-node-tests:_
https://github.com/input-output-hk/cardano-node-tests/issues/1265

File: cardano-api/src/Cardano/Api/Fees.hs (hunk 1 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Fees.hs]: I updated the comments in the code

File: cardano-api/src/Cardano/Api/Fees.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Fees.hs:1064]: Updated comments in the code

**Review by erikd — APPROVED** [review]

**Pull request merged**